File: R Programming/Tic_Tac_Toe.R
# Symbol chooser functions
symbol_verifier <- function() {
  cat("X or O? ")
  symbol <- toupper(readLines(con = con, n = 1))
  if (symbol %in% c("X", "O")) {
    return(symbol)
  }
  cat("Invalid symbol. Make sure to input either 'X' or 'O.'\n\n")
  symbol_verifier()
}

player_symbol_chooser <- function() {
  symbol <- symbol_verifier()
  cat("You successfully selected", symbol)
  return(symbol)
}

pc_symbol_chooser <- function(player_symbol) {
  if (player_symbol == "X") {
    return("O")
  } else {
    return("X")
  }
}

# Round declaration function
round_enunciator <- function(round_num) {
  central_line <- paste(strrep("#", 7), "Round", round_num, strrep("#", 8))
  cat("\n", strrep("#", 24), central_line, strrep("#", 24), "", sep = "\n")
  round_num <- round_num + 1
  return(round_num)
}

# Board setup and display functions
board_setup <- function() {
  board <- matrix(NA, nrow = 3, ncol = 3)
  return(board)
}

display_board <- function(board) {
  separator <- paste(strrep("~", 25), "\n")
  cat("Current board:\n")
  cat(separator)
  print(board)
  cat(separator)
}

# Coordinate validation functions
is_coord_valid <- function(coord) {
  if (coord %in% c("1", "2", "3")) {
    return(TRUE)
  } else {
    return(FALSE)
  }
}

is_coord_unnocupied <- function(board, row, column) {
  if (is.na(board[row, column])) {
    return(TRUE)
  } else {
    return(FALSE)
  }
}

# Player move prompt function

determine_placement <- function(board) {
  cat("Which row? ")
  row <- readLines(con = con, n = 1)
  cat("Which column? ")
  column <- readLines(con = con, n = 1)
  row_valid <- is_coord_valid(row)
  column_valid <- is_coord_valid(column)
  if (row_valid & column_valid) {
    row <- as.numeric(row)
    column <- as.numeric(column)
    coord_unnocupied <- is_coord_unnocupied(board, row, column)
    if (coord_unnocupied) {
      return(c(row, column))
    } else {
      cat("\nOccupied coordinate. Please select another one.\n\n")
    }
  } else {
    cat("\nInvalid coordinate. Make sure to input either 1, 2 or 3.\n\n")
  }
  determine_placement()
}

confirm_placement <- function(prompt, player_symbol, row, column, board) {
  cat("Place ", player_symbol, " at row ", row, ", column ", column, "? [y/n] ", sep = "")
  confirmation <- toupper(readLines(con = con, n = 1))
  if (confirmation == "Y") {
    board[row, column] <- player_symbol
    prompt <- FALSE
    return(c(prompt, board))
  } else if (confirmation == "N") {
    return(c(prompt, board))
  } else {
    cat("\nInvalid input. Make sure to input either 'y' or 'n'.\n\n")
  }
  confirm_placement(prompt, player_symbol, row, column, board)
}

prompt_move <- function(prompt, board, player_symbol) {
  coordinates <- determine_placement(board)
  row <- coordinates[1]
  column <- coordinates[2]
  res <- confirm_placement(prompt, player_symbol, row, column, board)
  prompt <- res[1]
  board <- matrix(res[2:10], nrow = 3, ncol = 3)
  if (prompt) {
    prompt_move(prompt, board, player_symbol)
  } else {
    return(board)
  }
}

player_move <- function(board, player_symbol) {
  prompt <- TRUE
  message("\nPlayer ", player_symbol, " turn\n")
  board <- prompt_move(prompt, board, player_symbol)
  cat("\nMove placed!\n\n")
  return(board)
}

# PC move prompt functions
check_line_for_opportunity <- function(line) {
  coord <- NA
  if (NA %in% line) {
    if (length(unique(line)) == 2) {
      if (sum(is.na(line)) == 1) {
        coord <- which(is.na(line))
        return(coord)
      }
    }
  }
  return(coord)
}

random_move <- function(board) {
  rand_row <- sample(1:3, 1)
  rand_col <- sample(1:3, 1)
  if (is_coord_unnocupied(board, rand_row, rand_col)) {
    move <- c(rand_row, rand_col)
    return(move)
  }
  random_move(board)
}

pc_move_choice <- function(board) {
  move <- NA
  for (i in 1:nrow(board)) {
    col <- check_line_for_opportunity(board[i, ])
    if (!is.na(col)) {
      move <- c(i, col)
      return(move)
    }
  }
  for (i in 1:ncol(board)) {
    row <- check_line_for_opportunity(board[, i])
    if (!is.na(row)) {
      move <- c(row, i)
      return(move)
    }
  }
  first_horizontal <- c(board[1, 1], board[2, 2], board[3, 3])
  coord <- check_line_for_opportunity(first_horizontal)
  if (!is.na(coord)) {
    move <- c(coord, coord)
    return(move)
  }
  second_horizontal <- c(board[1, 3], board[2, 2], board[3, 1])
  coord <- check_line_for_opportunity(second_horizontal)
  if (!is.na(coord)) {
    move <- c(coord, 4 - coord)
    return(move)
  }
  move <- random_move(board)
  return(move)
}

pc_move <- function(board, pc_symbol) {
  cat("\nPlayer", pc_symbol, "turn\n")
  move <- pc_move_choice(board)
  row <- move[1]
  col <- move[2]
  board[row, col] <- pc_symbol
  cat("PC Move registered!\n\n")
  return(board)
}

# Win conditions checker functions
check_line_for_win <- function(line, player_symbol) {
  if (!(NA %in% line)) {
    if (length(unique(line)) == 1) {
      if (unique(line) == player_symbol) {
        cat("You won!")
        return(TRUE)
      } else {
        cat("You lost!")
        return(TRUE)
      }
    }
  }
  return(FALSE)
}

check_win <- function(board, player_symbol, pc_symbol) {
  win <- FALSE
  for (i in 1:ncol(board)) {
    line_check <- check_line_for_win(board[, i], player_symbol)
    if (line_check) {
      return(TRUE)
    }
  }
  for (i in 1:nrow(board)) {
    line_check <- check_line_for_win(board[i, ], player_symbol)
    if (line_check) {
      return(TRUE)
    }
  }
  if (!is.na(board[2, 2])) {
    if (length(unique(c(board[1, 1], board[2, 2], board[3, 3]))) == 1 || length(unique(c(board[3, 1], board[2, 2], board[1, 3]))) == 1) {
      if (board[2, 2] == player_symbol) {
        cat("You won!")
        return(TRUE)
      } else if (board[2, 2] == pc_symbol) {
        cat("You lost!")
        return(TRUE)
      }
    } else if (!any(is.na(board))) {
      cat("Game over, no more space left!")
      return(TRUE)
    }
  }
  return(FALSE)
}

if (interactive()) {
  con <- stdin()
} else {
  con <- "stdin"
}

turn <- function(round, board, player_symbol, pc_symbol, win) {
  round <- round_enunciator(round)
  display_board(board)
  if (player_symbol == "X") {
    Sys.sleep(0.5)
    board <- player_move(board, player_symbol)
    Sys.sleep(0.5)
    display_board(board)
    win <- check_win(board, player_symbol, pc_symbol)
    if (win) {
      return(win)
    } else {
      Sys.sleep(0.5)
      board <- pc_move(board, pc_symbol)
    }
  } else {
    board <- pc_move(board, pc_symbol)
    Sys.sleep(0.5)
    display_board(board)
    win <- check_win(board, player_symbol, pc_symbol)
    if (win) {
      return(win)
    } else {
      Sys.sleep(0.5)
      board <- player_move(board, player_symbol)
    }
  }
  display_board(board)
  win <- check_win(board, player_symbol, pc_symbol)
  turn(round, board, player_symbol, pc_symbol, win)
}

replay <- function() {
  cat("\n\nWant to play again? [y/n] ")
  confirmation <- toupper(readLines(con = con, n = 1))
  if (confirmation == "Y") {
    cat("\n")
    game()
  } else if (confirmation == "N") {
    cat("\nThank you for playing! Until next time!")
    break
  } else {
    cat("\nInvalid input. Make sure to input either 'y' or 'n'.")
    replay()
  }
}

# Main function
game <- function() {
  player_symbol <- player_symbol_chooser()
  pc_symbol <- pc_symbol_chooser(player_symbol)
  board <- board_setup()
  round <- 1
  win <- FALSE
  turn(round, board, player_symbol, pc_symbol, win)
  replay()
}

game()

File: Python Programming/Tic_Tac_Toe.py
# Importing required libraries

import random
import time

# Symbol determination functions


def player_symbol_selector():
    symbol = input("X or O? ").upper()
    if symbol in ["X", "O"]:
        print("You successfully selected", symbol)
        return symbol
    else:
        print("Invalid symbol. Make sure to input either 'X' or 'O'.\n")
        return player_symbol_selector()


def pc_symbol_chooser(player_symbol):
    if player_symbol == "X":
        return "O"
    else:
        return "X"


# Round declaration function


def round_enunciator(round_num):
    central_line = f"{7*'#'} Round {str(round_num)} {8*'#'}"
    print(f"\n{24*'#'}\n{central_line}\n{24*'#'}")
    round_num += 1
    return round_num


# Defining board functionality


def board_setup():
    board = [["." for i in range(3)] for i in range(3)]
    return board


def display_board(board):
    separator = f"\n{24*'~'}\n"
    horizontal_divide = f"{5*'-'}+{5*'-'}+{+5*'-'}+{5*'-'}"
    print(
        f"\nCurrent board:\n{separator}\n{5*' '}|  1  |  2  |  3\n{horizontal_divide}"
    )
    for i in range(3):
        print(f"  {i+1}  |  {board[i][0]}  |  {board[i][1]}  |  {board[i][2]}")
        if i < 2:
            print(horizontal_divide)
        else:
            print(separator)


# Basic move verifications


def is_coord_valid(coord):
    if coord in ["1", "2", "3"]:
        return True
    else:
        return False


def is_coord_unnocupied(board, row, column):
    if board[row][column] == ".":
        return True
    else:
        return False


# Player move functions


def coord_prompter(coord_name):
    coord = input(f"Which {coord_name}? ")
    if is_coord_valid(coord):
        coord = int(coord) - 1
        return coord
    else:
        print("\nInvalid coordinate. Make sure to input either 1, 2 or 3.\n\n")
        return coord_prompter(coord_name)


def determine_placement(board):
    row = coord_prompter("row")
    column = coord_prompter("column")
    if is_coord_unnocupied(board, row, column):
        return row, column
    else:
        print("\nOccupied coordinate. Please select another one.\n\n")
        return determine_placement(board)


def confirm_placement(prompt, player_symbol, row, column, board):
    confirmation = input(
        f"Place {player_symbol} at row {row+1}, column {column+1}? [y/n] "
    ).upper()
    if confirmation == "Y":
        board[row][column] = player_symbol
        prompt = False
        return prompt, board
    elif confirmation == "N":
        return prompt, board
    else:
        print("\nInvalid input. Make sure to input either 'y' or 'n'.\n\n")
        return confirm_placement(prompt, player_symbol, row, column, board)


def prompt_move(prompt, board, player_symbol):
    row, column = determine_placement(board)
    prompt, board = confirm_placement(prompt, player_symbol, row, column, board)
    if prompt:
        prompt_move(prompt, board, player_symbol)
    else:
        return board


def player_move(board, player_symbol):
    prompt = True
    print(f"\nPlayer {player_symbol} turn\n")
    board = prompt_move(prompt, board, player_symbol)
    print("\nMove placed!\n\n")
    return board


# PC move functions


def check_line_for_opportunity(line):
    coord = None
    if "." in line:
        if len(list(set(line))) == 2:
            if line.count(".") == 1:
                coord = line.index(".")
                return coord
    return coord


def random_move(board):
    rand_row = random.choice(range(0, 3))
    rand_col = random.choice(range(0, 3))

    if is_coord_unnocupied(board, rand_row, rand_col):
        print(rand_row, rand_col)
        return rand_row, rand_col
    else:
        return random_move(board)


def pc_move_choice(board):
    for i in range(3):
        col = check_line_for_opportunity(board[i])
        if col is not None:
            return i, col
    for i in range(3):
        row = check_line_for_opportunity([board[0][i], board[1][i], board[2][i]])
        if row is not None:
            return row, i
    coord = check_line_for_opportunity([board[0][0], board[1][1], board[2][2]])
    if coord is not None:
        return coord, coord
    coord = check_line_for_opportunity([board[0][2], board[1][1], board[2][0]])
    if coord is not None:
        return coord, 3 - coord
    rand_row, rand_col = random_move(board)
    return rand_row, rand_col


def pc_move(board, pc_symbol):
    print(f"\nPlayer {pc_symbol} turn\n")
    row, col = pc_move_choice(board)
    board[row][col] = pc_symbol
    print("PC Move registered!\n\n")
    return board


# Win conditions checker functions


def check_line_for_win(line, player_symbol):
    if "." not in line:
        if len(list(set(line))) == 1:
            if line[0] == player_symbol:
                print("You won!")
                return True
            else:
                print("You lost!")
                return True
    return False


def check_win(board, player_symbol, pc_symbol):
    win = False
    for i in range(3):
        line_check = check_line_for_win(board[i], player_symbol)
        if line_check:
            return True
    for i in range(3):
        line_check = check_line_for_win(
            [board[0][i], board[1][i], board[2][i]], player_symbol
        )
        if line_check:
            return True
    if board[1][1] != ".":
        if (
            len(list(set([board[0][0], board[1][1], board[2][2]]))) == 1
            or len(list(set([board[0][2], board[1][1], board[2][0]]))) == 1
        ):
            if board[1][1] == player_symbol:
                print("You won!")
                return True
            elif board[1][1] == pc_symbol:
                print("You lost!")
                return True
        elif not any("." in line for line in board):
            print("Game over, no more space left!")
            return True
    return False


# Main game mechanisms


def turn(round_num, board, player_symbol, pc_symbol, win):
    round_num = round_enunciator(round_num)
    display_board(board)
    if player_symbol == "X":
        time.sleep(0.5)
        board = player_move(board, player_symbol)
        time.sleep(0.5)
        display_board(board)
        win = check_win(board, player_symbol, pc_symbol)
        if win:
            return win
        else:
            time.sleep(0.5)
            board = pc_move(board, pc_symbol)
    else:
        board = pc_move(board, pc_symbol)
        time.sleep(0.5)
        display_board(board)
        win = check_win(board, player_symbol, pc_symbol)
        if win:
            return win
        else:
            time.sleep(0.5)
            board = player_move(board, player_symbol)
    display_board(board)
    win = check_win(board, player_symbol, pc_symbol)
    if win:
        return win
    turn(round_num, board, player_symbol, pc_symbol, win)


def replay():
    confirmation = input("\n\nWant to play again? [y/n] ").upper()
    if confirmation == "Y":
        print("\n")
        game()
    elif confirmation == "N":
        print("\nThank you for playing! Until next time!")
    else:
        print("\nInvalid input. Make sure to input either 'y' or 'n'.")
        replay()


# Main function


def game():
    player_symbol = player_symbol_selector()
    pc_symbol = pc_symbol_chooser(player_symbol)
    board = board_setup()
    round_num = 1
    win = False
    turn(round_num, board, player_symbol, pc_symbol, win)
    replay()


game()